Purchases Page: Fix Bi-Yearly plan price showing incorrectly as yearly

spraveenitpro opened the pull request:
#### Description:

* Upon purchasing a Bi-yearly plan, when visiting the purchases page, it incorrectly shows as yearly.

#### Testing Instructions

1. Add a Business plan to cart and select to pay Bi annually i.e. every two years.
2. Complete the purchase.
3. Now visit the Purchase page at /purchases/subscriptions/:site/
4. Click the product to arrive at the purchase page.
5. You should correct see the `/ two years` like below:

![](https://d.pr/i/ht3blf+)

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #66025

spraveenitpro committed:
Fix Bi Yearly plan price showing incorrectly on Purchases page

sirbrillig reviewed (commented):
This works, but I'm curious _why_ it works.

I assume that the reason moving this block of assignments down solves the issue is that the `purchase.billPeriodLabel` block is adding the wrong text. If it's adding the wrong text, then `purchase.billPeriodLabel` must be `'per year'`. 🤔 That makes me wonder if `billPeriodLabel` is correct. Searching the codebase tells me two things: [it comes from an endpoint response](https://github.com/Automattic/wp-calypso/blob/ae44d99a72a24d54eb7274de7f653183306622ec/client/lib/purchases/assembler.ts#L11) with the key `bill_period_label`, and it is only used for this component (for now).

I wonder if a better solution is to modify the endpoint response so that `bill_period_label` has a "per two years" option. It looks like when it was originally implemented, no provision was made for biennial plans and the default was "per year" (sorry I can't link to the commit because it was before Phabricator but search for `get_bill_period_labels` and `BILL_PERIOD_2Y`).

If `bill_period_label` is modified to have "per two years", then here we could just modify the `purchase.billPeriodLabel` block to include a case for that value.

Also, by moving the block, you've made it take priority over monthly email products (not sure if that was intentional).

spraveenitpro commented:
> If it's adding the wrong text, then `purchase.billPeriodLabel` must be `'per year'`.

So `purchase.billPeriodLabel` is being checked in the context of [this PR ](https://github.com/Automattic/wp-calypso/pull/65006) which is targeting subscriptions for WooCommerce marketplace plugins from /plugins/woocommerce-subscriptions:

![](https://d.pr/i/axMqyI+)


>I wonder if a better solution is to modify the endpoint response so that bill_period_label has a "per two years" option. It looks like when it was originally implemented, no provision was made for biennial plans and the default was "per year" (sorry I can't link to the commit because it was before Phabricator but search for get_bill_period_labels and BILL_PERIOD_2Y).

Should we do this as a separate issue/PR?

>If bill_period_label is modified to have "per two years", then here we could just modify the purchase.billPeriodLabel block to include a case for that value.

That is true.

> by moving the block, you've made it take priority over monthly email products

You are right, I wanted to give the concerned block `plan && plan.term` priority over the `purchase.billPeriodLabel` block so I switched them up. Now, I have moved the `plan && plan.term` below the block `isEmailMonthly( purchase )`. Let me know if that is fine.

spraveenitpro committed:
Change priority of blocks by moving isEmailMonthly( purchase ) to have a higher priority

sirbrillig reviewed (approved):
Thank you for the explanation! Yes, I'm starting to see the shape of what happened here.

Originally we were using `plan.term` for plans, and some other early returns for specific purchases (that `isEmailMonthly()` check should probably be an early return). For Marketplace products, there was no `plan.term`, so the label used the default value, which is `'yearly'`. To fix this, the otherwise unused `bill_period_label` was available and was added for those products. However, it unintentionally affected _all_ products with that property, and that property is missing a biennial value both here and in the endpoint.

In this PR, you are giving `plan.term` (and `isEmailMonthly`) priority over `bill_period_label`, but that should be safe because it won't break the logic for products without `plan.term`.

What probably should really happen is that this component should be refactored to:

1. Use only early returns rather than reassigning `period` so that it's clear what each block's priority is.
2. Don't assume that the default term is "yearly"; if we cannot figure out the term in any other way, maybe display nothing at all and/or throw an error.
3. Only use `bill_period_label` if none of the other conditions is found (ie: only if `plan.term` does not exist, etc.).
4. Add unit tests to cover different products (marketplace, plans with different terms, one time purchases, transfers, included purchases, etc.) and verify their term is what is expected.
5. Add a `case` in the `bill_period_label` block for `'per two years'` (which will then require a separate backend change to add that string to the endpoint).

However, that can happen later. This fixes the bug in the meantime.

spraveenitpro commented:
### Additional Testing:

1. To confirm that we get the / two years when paying for two years: ![](https://d.pr/i/xTQFpF+)
2. To confirm that this change does not affect the /monthly display for the marketplace purchases as fixed at #65006  ![](https://d.pr/i/D0Ij4X+)

sirbrillig commented:
Looks good! Thanks for the thorough testing!
